### DISCUSSION TITLE
[@types/three] Object3D missing type for `this.material`

### DISCUSSION THREAD

**Discussion by johnnybenson** [DefinitelyTyped/DefinitelyTyped#57582] [Issues with a @types package]
Missing material type in Object3D

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/three/src/core/Object3D.d.ts

https://github.com/mrdoob/three.js/blob/dev/src/core/Object3D.js#L772

This may be more of an issue with the design of the class and ThreeJS than an issue with the type definitions.

```ts
// L772
if ( this.material !== undefined ) {
```

There is a reference to a member variable `material` in the class and no definition for it in the types.

I believe it should be this:

```ts
public material: Material | Material[] | undefined;
```

It's not clear that we should be typing member variables that are not initialized in the class constructor.

[a comment by an automated account is not shown]

**Comment by joshuaellis** [comment]
>This may be more of an issue with the design of the class and ThreeJS than an issue with the type definitions.

It's this, `Object3D`s cannot have a material. `Mesh`s can though. It's also only referenced in `toJSON` so I guess it's the design of the class to make that function a bit easier to write. You'd have to ask them for the specifics but I don't see this as an issue.

**Reply by Martin-Eckleben** [reply]
All the examples in vanilla JS do something like this:

```
const intersects = raycaster.intersectObjects( scene.children, false );
if ( intersects.length > 0 ) {
  if ( INTERSECTED != intersects[ 0 ].object ) {
    INTERSECTED = intersects[ 0 ].object;
    if ( INTERSECTED ) INTERSECTED.material.emissive.setHex( INTERSECTED.currentHex );
```

https://github.com/mrdoob/three.js/blob/master/examples/webgl_interactive_cubes.html

Which does not work with the current typings.

**Reply by Martin-Eckleben** [reply]
A workaround is:

```
if(this.INTERSECTED instanceof three.Mesh){
  this.INTERSECTED.material.emissive.setHex( INTERSECTED.currentHex );
}
else{
  console.error("intersected object is not a mesh")
}
```
